lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+	// estimate.
+	OptProvIndividualThresholdCertainty = 0.9
+
+	// OptProvSetThresholdStrictness describes the probability that the set of closest peers is actually further
+	// away then the calculated set threshold. Put differently, what is the probability that we are too strict and
+	// don't terminate the process early because we can't find any closer peers.
+	OptProvSetThresholdStrictness = 0.1
+
+	// OptProvReturnRatio corresponds to how many ADD_PROVIDER RPCs must have completed (regardless of success)
+	// before we return to the user. The ratio of 0.75 equals 15 RPC as it is based on the Kademlia bucket size.
+	OptProvReturnRatio = 0.75
+)
+
+type addProviderRPCState int
+
+const (
+	Sent addProviderRPCState = iota + 1
+	Success
+	Failure
+)
+
+type estimatorState struct {
+	// context for all ADD_PROVIDER RPCs
+	putCtx context.Context
+
+	// reference to the DHT
+	dht *IpfsDHT
+
+	// the most recent network size estimation
+	networkSize float64
+
+	// a channel indicating when an ADD_PROVIDER RPC completed (successful or not)
+	doneChan chan struct{}
+
+	// tracks which peers we have stored the provider records with
+	peerStatesLk sync.RWMutex
+	peerStates   map[peer.ID]addProviderRPCState
+
+	// the key to provide
+	key string
+
+	// the key to provide transformed into the Kademlia key space
+	ksKey ks.Key
+
+	// distance threshold for individual peers. If peers are closer than this number we store
+	// the provider records right away.
+	individualThreshold float64
+
+	// distance threshold for the set of bucketSize closest peers. If the average distance of the bucketSize
+	// closest peers is below this number we stop the DHT walk and store the remaining provider records.
+	// "remaining" because we have likely already stored some on peers that were below the individualThreshold.
+	setThreshold float64
+
+	// number of completed (regardless of success) ADD_PROVIDER RPCs before we return control back to the user.
+	returnThreshold int
+
+	// putProvDone counts the ADD_PROVIDER RPCs that have completed (successful and unsuccessful)
+	putProvDone atomic.Int32
+}
+
+func (dht *IpfsDHT) newEstimatorState(ctx context.Context, key string) (*estimatorState, error) {
+	// get network size and err out if there is no reasonable estimate
+	networkSize, err := dht.nsEstimator.NetworkSize()
+	if err != nil {
+		return nil, err
+	}
+
+	individualThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize), 1-OptProvIndividualThresholdCertainty) / networkSize
+	setThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize)/2.0+1, 1-OptProvSetThresholdStrictness) / networkSize
+	returnThreshold := int(math.Ceil(float64(dht.bucketSize) * OptProvReturnRatio))
+
+	return &estimatorState{
+		putCtx:              ctx,
+		dht:                 dht,
+		key:                 key,
+		doneChan:            make(chan struct{}, returnThreshold), // buffered channel to not miss events
+		ksKey:               ks.XORKeySpace.Key([]byte(key)),
```

Comment:
***`[]byte(key)`***
![wall e](https://user-images.githubusercontent.com/24391983/214761328-c887ed36-380a-4c67-95f9-3a0a0564a67c.png)
But that linked to the fact the keys are not stored as `[32]byte`, so not your fault. (it's a much bigger refactor to fix this, not worth fixing now)

Reply:
🤷